_drafts/making-backend-scalable.markdown

```diff
@@ -0,0 +1,123 @@
+---
+layout: post
+title: "Making Autolab's Backend Scalable"
+author: Ilter Canberk
+
+tags: [tango, redis, scalability]
+
+---
+
+Tango is the back-end service that receives grading jobs from Autolab's front-end, puts them in a job queue, and assigns them to available containers for grading. Tango came across larger work-loads and huge files as Autolab got more popular this semester.
+
+
+![Tango's initial Architecture]({{site-url}}/assets/redis1.png)
+_Diagram 1: Initial Tango architecture with in-memory job queue_
+
+
+The existing monolithic model with a single process and an in-memory queue limited the scalability and was prone to crashes that caused the jobs to disappear. Therefore we decided to improve our architecture by switching to multi-process model with a persistent queue.
```

Comment:
The existing monolithic model with a single process and an in-memory job queue was posing multiple problems:

* Running a web server that handled large file uploads concurrently along with a job manager that dispatched jobs from the job queue to new threads made the process prone to crashes.
* If this single Tango process crashed, all the jobs stored in the in-memory queue are lost.
* All the containers running a job are suddenly left in a so-called limbo state because they are no longer managed by any Tango process.